### ISSUE TITLE
@provide support

### ISSUE THREAD

**Issue by remojansen** [inversify/InversifyJS#190]
## Expected Behavior

Developers want a new decorator `@provide` which generates bindings for us. The `@provide` decorator needs access to a kernel instance;

``` ts
let kernel = new Kernel();
let provide = makeProvideDecorator(kernel);
```

The `@provide` decorator will have a very similar behavior to the `kernel.bind` fluent API:

``` ts
// AKA kernel.bind<any>(MyService1).to(MyService1).inSingletonScope()
@provide().inSingletonScope().done()
class MyService1 {}

// AKA kernel.bind<any>(MyService2).to(MyService2).inTransientScope()
@provide().inTransientScope().done()
class MyService2 {}

// AKA kernel.bind<any>(Katana).to(Katana).whenTargetTagged("canThrow", false)
@provide().whenTargetTagged("canThrow", false).done()
class Katana {}

// AKA kernel.bind<any>(Shuriken).to(Shuriken).whenTargetTagged("canThrow", true)
@provide().whenTargetTagged("canThrow", true).done()
class Shuriken {}
```

When the `@provide` decorator is used the `@injectable` decorator will not be mandatory.

We will also provide documented recipes for developers to explain how to create aliases:

``` ts
let provideSingleton = provide().inSingletonScope().done();
let provideTransient = provide().inTransientScope().done();
let provideThrowable = provide().whenTargetTagged("canThrow", true).done();
```

This will make the API less verbose:

``` ts
@provideSingleton
class Katana {}

@provideTransient
class Katana {}

@provideThrowable
class Shuriken {}
```
## Current Behavior

Declaration of bindings via `@provide` is not currently supported.
## Possible Solution

We can wrap the `kernel.bind` fluent syntax so it includes the `done` method. The done method must return a class decorator. The class decorator will invoke the `kernel.bind` method.
### Context

This feature was proposed by @otbe on #187. 

This might be developed as an optional plugin:

``` ts
import makeProvideDecorator from "inversify-utils";
import { Kernel } from "inversify";

let kernel = new Kernel();
let provide = makeProvideDecorator(kernel);

let provideSingleton = provide().inSingletonScope().done();
let provideTransient = provide().inTransientScope().done();
let provideThrowable = provide().whenTargetTagged("canThrow", true).done();

export { provide, provideSingleton, provideTransient, provideThrowable };
```

``` ts
import { 
    provide, 
    provideSingleton, 
    provideTransient, 
    provideThrowable 
} from "./ioc/decorators";

@provideSingleton
class Katana {}

@provideTransient
class Katana {}

@provideThrowable
class Shuriken {}

@provide().whenTargetTagged("canThrow", true).done();
class Katana {}
```

**Comment by remojansen** [comment]
I'm going to release this as an optional npm package. I'm going to be working on  another repo: https://github.com/inversify/inversify-binding-decorators

**Comment by remojansen** [comment]
I just released [inversify-binding-decorators@1.0.0-beta.1](https://github.com/inversify/inversify-binding-decorators/releases) so I'm closing this issue.

**Comment by ryall** [comment]
Shame this is included separately. I completely missed it up until now because I never found this issue. Maybe a reference in the core docs would be helpful.
